sync/awsdatabricks.py

```diff
@@ -706,6 +710,11 @@ def _s3_contents_have_all_rollover_logs(contents: list[dict], run_end_time_secon
     )
 
 
+def event_log_poll_duration_seconds():
+    """Convenience function to aid testing"""
+    return 15
```

Comment:
Also for testing, so that I can easily set this to `0` to avoid needing to wait forever when testing the polling logic

Reply:
> I think it better to check for contents before sleeping in _get_eventlog. ...  or maybe you can put the check in the loop statement....

I agree! And yeah, we have the check in the loop - we skip the `sleep` on `num_attempts == 0`, so the first call is made immediately. If we are satisfied with the contents in the bucket, then we just move on without ever calling `sleep`!

I still need to keep this here, though, because I have a test case that _is_ testing the polling scenario, and I don't want that test to take like a minute to run.


> In my limited manual testing the event log was immediately available

So, there is a little bit of complexity here. For longer-running jobs, there will always be SOME eventlog data available. The only question is if ALL of the eventlog data has been flushed. Databricks should flush all the data upon cluster termination, but sometimes we actually catch the cluster while it's still RUNNING and not yet TERMINATED. I linked this in a few comments in code, but they have some documentation here about this - https://docs.databricks.com/clusters/configure.html#cluster-log-delivery-1

So without any polling mechanism, we can fool ourselves into thinking that we have everything we need, when in reality we are missing a bit of the event log data. I actually caught this because I was testing using a very _short_ job (about ~2 minutes to execute), and was frequently getting a "No eventlog found at location" error (even though it _was_ there when I looked manually)

Reply:
I do wait on the cluster when running a one-off job (with either project or prediction configuration applied): https://github.com/synccomputingcode/syncsparkpy/blob/2919c60058acd73889f767d97593ec87dec62d69/sync/awsdatabricks.py#L557

One alternative that may be worth considering is adding a wait on the cluster in addition to the job run when we're just executing an existing job (/2.1/jobs/run-now).

Reply:
yeah, and I was testing with `run-job`, which should be calling into this. I'm glad you pointed this out, though, because I'd missed it - it seems to me that `existing_cluster_id` only exists if you are running a Job/Task on a cluster that was already provisioned. If instead a new cluster is launched when you start a Job run, `existing_cluster_id` will not be present -

existing_cluster_id | string If  existing_cluster_id, the ID of an existing cluster that is used for all  runs of this task. When running tasks on an existing cluster, you may  need to manually restart the cluster if it stops responding. We suggest  running jobs on new clusters for greater reliability.
-- | --

https://docs.databricks.com/dev-tools/api/latest/jobs.html#operation/JobsRunNow

I don't necessarily want to block on the cluster terminating _here_, because we do want to make the EC2 `describe_instances` call as soon as possible, but that could be a good alternative to polling S3, for sure

Reply:
Even still, we would likely still want to keep the polling for the `LastModified` time, though. Since we know now that users can configure log destinations in regions other than the one their DBX workspace is in, that additional cross-region latency will likely cause consistency issues for us in some cases

Reply:
Yeah, just ran a quick test. Looks like for runs based on an existing job (/2.1/jobs/run-now) the cluster ID will be at .cluster_instance.cluster_id for tasks that are at the PENDING or later lifecycle stage...